src/rapid_pro_to_engagement_db/rapid_pro_to_engagement_db.py

```diff
@@ -226,60 +227,82 @@ def sync_rapid_pro_to_engagement_db(rapid_pro, engagement_db, uuid_table, rapid_
+                    msg = Message(
+                        participant_uuid=participant_uuid,
+                        text=rapid_pro_result.input,  # Raw text received from a participant
+                        timestamp=rapid_pro_result.time,  # Time at which Rapid Pro processed this message in the flow.
+                        direction=MessageDirections.IN,
+                        channel_operator=URNCleaner.clean_operator(contact_urn),
+                        status=MessageStatuses.LIVE,
+                        dataset=config.engagement_db_dataset,
+                        labels=[],
+                        origin=MessageOrigin(
+                            origin_id=f"rapid_pro.workspace_{workspace_uuid}.flow_{flow_id}.run_{run.id}.result_{rapid_pro_result.name}",
+                            origin_type="rapid_pro"
+                        )
+                    )
+                    message_origin_details = {
+                        "rapid_pro_workspace": workspace_name,
+                        "run_id": run.id,
+                        "flow_id": flow_id,
+                        "flow_name": flow_name,
+                        "run_value": rapid_pro_result.serialize()
+                    }
+                    sync_event = _ensure_engagement_db_has_message(engagement_db, msg, message_origin_details, dry_run)
+                    sync_stats.add_event(sync_event)
+                dataset_to_sync_stats[f"{flow_name}.{config.flow_result_field}"].add_stats(sync_stats)
+            
             # Update the cache so we know not to check this run again in this flow + result field context.
+            # TODO: Update the cache if we've read the last run, or the next run's last modified timestamp is greater
+            # than the run we are currently syncing.
             if cache is not None and not dry_run:
-                cache.set_latest_run_timestamp(flow_id, flow_config.flow_result_field, run.modified_on)
-
-        dataset_to_sync_stats[f"{flow_config.flow_name}.{flow_config.flow_result_field}"] = sync_stats
+                cache.set_latest_run_timestamp(flow_id, run.modified_on)
 
-    # Log the summaries of actions taken for each dataset then for all datasets combined.
-    all_sync_stats = RapidProToEngagementDBSyncStats()
-    for flow_config in rapid_pro_config.flow_result_configurations:
-        result_field = f"{flow_config.flow_name}.{flow_config.flow_result_field}"
-        log.info(f"Summary of actions for flow result field '{result_field}':")
-        dataset_to_sync_stats[result_field].print_summary()
-        all_sync_stats.add_stats(dataset_to_sync_stats[result_field])
+        for _ in runs:
+            stats.add_event(RapidProSyncEvents.READ_RUN_FROM_RAPID_PRO)
+        flow_stats[flow_name] = stats
+
+    # Log the summaries of actions taken for each flow and each dataset then for all flows and datasets combined.
+    all_flow_stats = FlowStats()
+    all_sync_stats = FlowResultToEngagementDBSyncStats()
```

Comment:
This is a really nice solution - we should consider something similar in the CSV sync

Reply:
Thanks. I'll work on it.